desc/protoparse/ast/walk.go

```diff
@@ -0,0 +1,493 @@
+	case *ServiceNode:
+		if v.VisitServiceNode != nil {
+			matched = true
+			ok, next = v.VisitServiceNode(n)
+		}
+	case *RPCNode:
+		if v.VisitRPCNode != nil {
+			matched = true
+			ok, next = v.VisitRPCNode(n)
+		}
+	case *RPCTypeNode:
+		if v.VisitRPCTypeNode != nil {
+			matched = true
+			ok, next = v.VisitRPCTypeNode(n)
+		}
+	case *IdentNode:
+		if v.VisitIdentNode != nil {
+			matched = true
+			ok, next = v.VisitIdentNode(n)
+		}
+	case *CompoundIdentNode:
+		if v.VisitCompoundIdentNode != nil {
+			matched = true
+			ok, next = v.VisitCompoundIdentNode(n)
+		}
+	case *StringLiteralNode:
+		if v.VisitStringLiteralNode != nil {
+			matched = true
+			ok, next = v.VisitStringLiteralNode(n)
+		}
+	case *CompoundStringListeralNode:
+		if v.VisitCompoundStringLiteralNode != nil {
+			matched = true
+			ok, next = v.VisitCompoundStringLiteralNode(n)
+		}
+	case *UintLiteralNode:
+		if v.VisitUintLiteralNode != nil {
+			matched = true
+			ok, next = v.VisitUintLiteralNode(n)
+		}
+	case *PositiveUintLiteralNode:
+		if v.VisitPositiveUintLiteralNode != nil {
+			matched = true
+			ok, next = v.VisitPositiveUintLiteralNode(n)
+		}
+	case *NegativeIntLiteralNode:
+		if v.VisitNegativeIntLiteralNode != nil {
+			matched = true
+			ok, next = v.VisitNegativeIntLiteralNode(n)
+		}
+	case *FloatLiteralNode:
+		if v.VisitFloatLiteralNode != nil {
+			matched = true
+			ok, next = v.VisitFloatLiteralNode(n)
+		}
+	case *SpecialFloatLiteralNode:
+		if v.VisitSpecialFloatLiteralNode != nil {
+			matched = true
+			ok, next = v.VisitSpecialFloatLiteralNode(n)
+		}
+	case *SignedFloatLiteralNode:
+		if v.VisitSignedFloatLiteralNode != nil {
+			matched = true
+			ok, next = v.VisitSignedFloatLiteralNode(n)
+		}
+	case *BoolLiteralNode:
+		if v.VisitBoolLiteralNode != nil {
+			matched = true
+			ok, next = v.VisitBoolLiteralNode(n)
+		}
+	case *SliceLiteralNode:
+		if v.VisitSliceLiteralNode != nil {
+			matched = true
+			ok, next = v.VisitSliceLiteralNode(n)
+		}
+	case *AggregateLiteralNode:
+		if v.VisitAggregateLiteralNode != nil {
+			matched = true
+			ok, next = v.VisitAggregateLiteralNode(n)
+		}
+	case *KeywordNode:
+		if v.VisitKeywordNode != nil {
+			matched = true
+			ok, next = v.VisitKeywordNode(n)
+		}
+	case *RuneNode:
+		if v.VisitRuneNode != nil {
+			matched = true
+			ok, next = v.VisitRuneNode(n)
+		}
+	case *EmptyDeclNode:
+		if v.VisitEmptyDeclNode != nil {
+			matched = true
+			ok, next = v.VisitEmptyDeclNode(n)
+		}
+	}
+
+	if !matched {
+		// Visitor provided no concrete type visit function, so
+		// check interface types. We do this in several passes
+		// to provide "priority" for matched interfaces for nodes
+		// that actually implement more than one interface.
+		//
+		// For example, StringLiteralNode implements both
+		// StringValueNode and ValueNode. Both cases could match
+		// so the first case is what would match. So if we want
+		// to test against either, they need to be in different
+		// switch statements.
```

Comment:
This could be guarded against with an explicit ordering, no? Type switches match in order, so we can just order things by preference, e.g.

```go
switch n := n.(type) {
case IntValueNode:
	if v.VisitIntValueNode != nil {
		matched = true
		ok, next = v.VisitIntValueNode(n)
	}
case FloatValueNode:
	if v.VisitFloatValueNode != nil {
		matched = true
		ok, next = v.VisitFloatValueNode(n)
	}
}
```

Ref: https://golang.org/ref/spec#TypeSwitchStmt.

Reply:
The problem is that a `*UintLiteralNode` matches both, so would _always_ all into the first case -- for `IntValueNode`.

But if the visitor has no `VisitIntValueNode`, we need to instead invoke any other matching case (such as `VisitFloatValueNode, `ValueValueNode`, VisitTerminalNode`, or `VisitNode`.

I could also make this more explicit by _only_ having cases for each concrete type and then use `if/else` or another `switch` to check each possible matching function. But that felt more error-prone (easy to make a node implement an interface but then forget to update the case statement in this function).

I guess another way to skin this case could be code gen: a code generator could load up all possible node types and compute what interfaces they satisfy, and then I could _generate_ the body of this function. In that case, the code would probably instead have something like this:

```go
case *UintLiteralNode:
    if v.VisitUintLiteralNode != nil {
    } else if v.VisitIntValueNode != nil {
    } else if v.VisitFloatValueNode != nil {
    } else if v.VisitValueNode != nil {
    } else if v.VisitTerminalNode != nil {
    } else if v.VisitNode != nil {
    }
```

And it would not be error-prone if it were generated instead of hand-written... But I would need a way to control the preference order for generating the `if` conditions, so they come out in the correct order. That is mostly easy (can test when one interface implements another and put the broadest one last), but that doesn't work for the two cases described in comments: nodes that implement both `IntValueNode` and `FloatValueNode` (of which there is only one) and nodes that implement both `FieldDeclNode` and `MessageDeclNode` (of which there are two).

Reply:
I will probably need to write a code generator to make sure this is always in sync and correct. I caught a missing case for example (`*MessageFieldNode`, for entries in a message literal, as can be found in complex custom options).